### ISSUE TITLE
Technical Report: Task-Specific Fine-Tuning of e5-small-v2 as a Model Compression Strategy

### ISSUE THREAD

**Issue by craigtrim** [craigtrim/edu-alias-mapper-e5#2]
## Abstract

This issue documents the theoretical basis, implementation decisions, and evaluation methodology for compressing the fine-tuned \`intfloat/e5-large-v2\` retrieval model (~1.3 GB, 1024-dim) into a deployable artifact using \`intfloat/e5-small-v2\` (~90 MB, 384-dim). The approach is **direct task-specific fine-tuning**, not knowledge distillation in the strict sense. This distinction carries significant theoretical implications that are documented below, as they bear on the interpretation of evaluation results and inform contingency paths if quality gates fail.

---

### 1.1 Knowledge Distillation (Hinton et al., 2015)

Classical knowledge distillation trains a student network to match the **output distribution** of a teacher network, rather than training on hard labels from the original dataset. The central insight is that a teacher's softened output probabilities (at temperature T > 1) carry **dark knowledge** — relational information about inter-class similarities that is absent from one-hot ground-truth labels.

For generative language models this takes the form of KL divergence between teacher and student logit distributions over the vocabulary:

```
L_KD = KL( softmax(z_T / T) || softmax(z_S / T) )
```

where z_T and z_S are teacher and student logits respectively and T is the temperature hyperparameter.

### 1.2 Embedding Distillation (Sanh et al., 2019; Reimers & Gurevych, 2020)

For sentence embedding models there are no vocabulary logits. The analogous objective minimises the mean squared error between teacher and student embedding vectors:

```
L_MSE = (1/n) * Σ_i || φ_T(x_i) - φ_S(x_i) ||²
```

where φ_T and φ_S are the teacher and student embedding functions. This objective explicitly encourages the student to reproduce the **geometric structure** of the teacher's embedding space — pairwise distances, nearest-neighbour topology, and cluster geometry — not merely to match the sign of similarity for known pairs.

### 1.3 What Was Actually Implemented Here

The approach implemented in \`scripts/train_small_model.py\` is **neither of the above**. It is direct supervised fine-tuning of the student base model on the original (alias, label) training pairs, using the same contrastive loss objective that was used to train the teacher. The teacher model plays no role during training.

The more precise term is **task-specific fine-tuning of a smaller pretrained model**, or in the model compression literature, **independent fine-tuning** of the student (as opposed to distillation-guided fine-tuning).

This distinction matters because:

1. The student's embedding space is shaped entirely by the base pretraining of \`e5-small-v2\` and the contrastive fine-tuning signal from the alias-label pairs. It is **not** constrained to match the teacher's learned geometry.
2. Any accuracy gap relative to the teacher reflects a combination of (a) reduced model capacity and (b) the absence of the teacher's implicit knowledge. Embedding distillation would address (b) without changing (a).
3. If quality gates fail, the correct next step is embedding distillation via MSELoss — not re-running this pipeline at higher epochs or with a larger batch.

---

### 2.1 Teacher: intfloat/e5-large-v2

- Architecture: Transformer encoder, 24 layers, 16 attention heads
- Hidden dimension: 1024
- Parameter count: ~335 million
- Output: Mean-pooled, L2-normalised 1024-dimensional dense vector
- Disk footprint (fine-tuned checkpoint): ~1.3 GB

### 2.2 Student Base: intfloat/e5-small-v2

- Architecture: Transformer encoder, 12 layers, 12 attention heads
- Hidden dimension: 384
- Parameter count: ~33 million (approximately 10× fewer than teacher)
- Output: Mean-pooled, L2-normalised 384-dimensional dense vector
- Disk footprint (base weights): ~90 MB

The representational capacity reduction is substantial. A 384-dimensional embedding space can express at most 384 orthogonal directions of semantic variation, versus 1024 for the large model. For a constrained-domain task (institutional alias mapping over a fixed vocabulary of ~tens of thousands of entities), this compression may be lossless in practice — the effective intrinsic dimensionality of the alias-label semantic manifold is unlikely to approach either limit.

---

### 3.1 Formal Definition

Given a batch B = {(a_i, p_i)}_{i=1}^{N} of (anchor, positive) pairs, MultipleNegativesRankingLoss (MNRL) treats all other positives in the batch as hard negatives for each anchor. The loss is:

```
L_MNRL = -(1/N) * Σ_i log [ exp(sim(a_i, p_i) / τ) / Σ_{j=1}^{N} exp(sim(a_i, p_j) / τ) ]
```

where sim(·,·) denotes cosine similarity and τ is a (learned or fixed) temperature parameter.

This objective is structurally equivalent to **NT-Xent** (Normalized Temperature-scaled Cross Entropy Loss) from SimCLR (Chen et al., 2020), and to the InfoNCE loss (Oord et al., 2018). It is also sometimes called **in-batch negatives** or **batch softmax loss** in the information retrieval literature.

### 3.2 Properties Relevant to This Task

**Implicit negative mining:** With batch size N=64, each anchor sees 63 in-batch negatives. The quality of these negatives depends on the batch composition. Since the dataset contains institutional aliases from a heterogeneous global corpus (DBpedia schools), random batching is unlikely to produce many false negatives (i.e., batches where a true positive appears as a negative), which is the principal failure mode of this loss.

**Asymmetric training signal:** MNRL only requires positive pairs — no explicit negatives are needed. This suits the alias-label dataset, which is inherently structured as matched pairs without curated hard negatives.

**Contrast with TripletLoss and CosineSimilarityLoss:** TripletLoss requires explicit (anchor, positive, negative) triples; CosineSimilarityLoss optimises pairwise cosine distance with scalar labels. MNRL is generally preferred for retrieval tasks because it optimises ranking directly rather than a proxy pairwise similarity objective.

---

### 4.1 Dimension Incompatibility

FAISS \`IndexFlatIP\` encodes the vector dimension d at index construction time and validates it on every \`search()\` call. An index built at d=1024 (teacher) will raise a runtime exception if queried with a d=384 vector (student). There is no migration path — the index must be rebuilt from scratch at the new dimension.

This script writes to separate index paths:
- \`faiss/labels_small.index\` (d=384)
- \`faiss/aliases_small.index\` (d=384)

preserving the teacher's indexes at:
- \`faiss/labels.index\` (d=1024)
- \`faiss/aliases.index\` (d=1024)

### 4.2 Search Complexity

\`IndexFlatIP\` performs exhaustive inner-product search in O(n · d) time per query. The dimension reduction from 1024 to 384 yields a theoretical **2.67× speedup** in search time, independent of index size. At the corpus sizes in this project (tens of thousands of vectors) this is unlikely to be a practical bottleneck in either case, but it is a non-trivial latency improvement for high-throughput Lambda invocations.

### 4.3 Index File Size

Each vector stored in an \`IndexFlatIP\` index occupies \`d * sizeof(float32)\` = \`d * 4\` bytes. For a corpus of n vectors:

- Teacher index: n * 1024 * 4 bytes = ~4.1 KB per vector
- Student index: n * 384 * 4 bytes = ~1.5 KB per vector

A corpus of 50,000 unique labels produces:
- Teacher: ~200 MB per index
- Student: ~75 MB per index

---

### 5.1 Ground-Truth Definition

Top-1 retrieval accuracy is defined over (alias, label) pairs from the dataset. For each alias query, the model retrieves the highest-scoring label from the full label corpus via FAISS. A retrieval is correct if and only if the retrieved label string exactly matches the ground-truth label string paired with the alias in the dataset. This is a strict criterion — partial matches, synonyms, and alternative forms are all counted as incorrect.

### 5.2 Quality Gates

Three independent gates are evaluated over a random sample of held-out alias-label pairs (default n=1000, configurable via \`--sample\`):

**Gate A — Accuracy Gap (relative to teacher)**
```
student_accuracy >= teacher_accuracy - 0.05
```
Rationale: A degradation of more than 5 percentage points relative to the teacher indicates that the capacity reduction is causing material quality loss, not merely expected variance.

**Gate B — Retrieval Agreement**
```
agreement_rate >= 0.88
```
where agreement_rate = P(student_top1 == teacher_top1). This metric captures whether the two models have learned compatible retrieval geometries. A high agreement rate with a low absolute accuracy (Gate C failure) would indicate that both models are wrong in the same way — a signal to revisit the training data rather than the model architecture.

**Gate C — Standalone Minimum Floor**
```
student_accuracy >= 0.85
```
An absolute lower bound independent of teacher performance. Guards against a degenerate case where the teacher is itself mediocre and Gate A passes by proximity.

### 5.3 Limitation: Evaluation Set Leakage

The evaluation pairs are drawn from the same dataset used for training. There is no held-out test split in the current pipeline. This means the accuracy metrics measure **retrieval on the training distribution**, not generalisation to unseen aliases. For deployment purposes this is acceptable if the deployment distribution is close to the training distribution (i.e., the alias vocabulary is largely closed). For a more rigorous evaluation, a held-out split should be constructed before training begins.

---

## 6. Theoretical Justification for Direct Fine-Tuning Over Distillation

The primary argument against direct fine-tuning (and for distillation) is that the teacher's fine-tuned embedding space encodes pairwise relational structure that is not recoverable from the raw positive pairs alone. Specifically, the teacher has learned **soft negatives** — the degree to which "UdeG" is similar to "University of Guadalajara" versus "Universidad de Guanajuato" — through the geometric arrangement of its embedding space. MNRL on raw pairs communicates only binary signal: each positive pair should be closer than in-batch random pairs.

However, direct fine-tuning is justified as the first attempt for the following reasons:

1. **Domain constraint:** The alias-label space is highly structured and relatively low-entropy. Many aliases are unambiguous abbreviations or transliterations that a well-pretrained small model should resolve correctly without teacher guidance.

2. **Base model pretraining:** \`e5-small-v2\` was pretrained using weak supervision on a large multilingual corpus with contrastive objectives (Wang et al., 2022). Its embedding geometry already encodes substantial institutional semantic structure before fine-tuning begins.

3. **Empirical precedent:** On MTEB retrieval benchmarks, \`e5-small-v2\` scores approximately 49.0 nDCG@10 versus \`e5-large-v2\`'s 56.0 — a ~7-point gap on diverse open-domain retrieval. For a closed-domain task with a fixed vocabulary, this gap is expected to be smaller.

4. **Falsifiability:** If Gate A fails (accuracy gap > 5pp), this hypothesis is falsified and embedding distillation via \`sentence_transformers.losses.MSELoss\` is the prescribed next step.

---

## 7. Contingency Path (if Gates Fail)

**Step 1: Embedding distillation**

Replace the training objective in \`train_small_model.py\` with \`MSELoss\` from \`sentence_transformers.losses\`. Precompute and cache teacher embeddings for the full corpus (avoids repeated teacher forward passes). Train the student to minimise coordinate-wise MSE between its outputs and the frozen teacher's outputs.

**Step 2: e5-base-v2**

If distillation with e5-small-v2 fails Gate C, escalate to \`intfloat/e5-base-v2\` (~440 MB, 768-dim). This halves the accuracy gap relative to large while remaining substantially smaller. Re-evaluate Lambda cold-start latency — 440 MB may still be within the 10-second budget depending on Lambda memory allocation.

**Step 3: ONNX + quantization**

Export the fine-tuned student to ONNX format and apply INT8 static quantization. This typically yields 30–50% latency reduction on CPU with minimal accuracy loss for Transformer encoders. The pooling and normalisation layers must be reimplemented manually or verified against the ONNX export — this is the most common failure point in this path.

---

## References

- Hinton, G., Vinyals, O., & Dean, J. (2015). Distilling the Knowledge in a Neural Network. *arXiv:1503.02531*
- Sanh, V., Debut, L., Chaumond, J., & Wolf, T. (2019). DistilBERT, a distilled version of BERT. *arXiv:1910.01108*
- Reimers, N., & Gurevych, I. (2020). Making Monolingual Sentence Embeddings Multilingual using Knowledge Distillation. *EMNLP 2020*
- Chen, T., Kornblith, S., Norouzi, M., & Hinton, G. (2020). A Simple Framework for Contrastive Self-Supervised Learning. *ICML 2020*
- Oord, A., Li, Y., & Vinyals, O. (2018). Representation Learning with Contrastive Predictive Coding. *arXiv:1807.03748*
- Wang, L., Yang, N., Huang, X., Jiao, B., Yang, L., Jiang, D., Majumder, R., & Wei, F. (2022). Text Embeddings by Weakly-Supervised Contrastive Pre-training. *arXiv:2212.03533*
